### PR TITLE
Feature: Add hooks for layout "pre" and "post"

### PR DESCRIPTION
**Pull request by KoolADE85**
This PR adds new renderer hooks for users to write code that reads & updates the page layout.
In the same style as `request_pre` and `request_post`, there is now a `layout_pre` and a `layout_post`.

## Contributor Checklist

- [x] I have run the tests locally and they passed. (refer to testing section in [contributing](https://github.com/plotly/dash/blob/master/CONTRIBUTING.md))
- [x] I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR

### optionals

- [x] I have added entry in the `CHANGELOG.md`
- [ ] If this PR needs a follow-up in **dash docs**, **community thread**, I have mentioned the relevant URLS as follows
    -  [ ] this GitHub [#PR number]() updates the dash docs
    -  [ ] here is the show and tell thread in Plotly Dash community

### CODE REVIEW

File: tests/integration/renderer/test_request_hooks.py [hunk 1]
```diff
@@ -295,3 +295,46 @@ def wrap(*args, **kwargs):
...
+            {%favicon%}
+            {%css%}
+        </head>
+        <body>
+            {%app_entry%}
+            <footer>
+                {%config%}
+                {%scripts%}
+                <script id="_dash-renderer" type"application/json">
+                    const renderer = new DashRenderer({
```
Comment (T4rk1n) [inline tests/integration/renderer/test_request_hooks.py:317]: This api usage is not something I think we should expand upon without improvements, it's really hard to use and has boilerplate not required for the feature.

I suggest we add parameter(s) that get them in the main index without the need to override the index string.

Comment (KoolADE85) [reply tests/integration/renderer/test_request_hooks.py:317]: oh sure, I was just following the same style as the other tests in this file.
I agree it's odd to overwrite the entire index_string. I can also do this instead:

```
app.renderer = """
    new DashRenderer({
        ... 
    });
"""
```

which would resemble what I'm actually doing in EnterpriseDash

Comment (T4rk1n) [reply tests/integration/renderer/test_request_hooks.py]: No I meant we should add a parameter to init like:
```python
app = Dash(hooks={"pre_layout": "() => null"})
```
This way you don't need to override the index values and the boilerplate is minimal.

Comment (KoolADE85) [reply tests/integration/renderer/test_request_hooks.py]: As discussed, I've added the ability to pass these hooks as an argument to the constructor.

File: dash/dash.py [hunk 2]
```diff
@@ -373,6 +375,7 @@ def __init__(  # pylint: disable=too-many-statements
         long_callback_manager=None,
         background_callback_manager=None,
         add_log_handler=True,
+        hooks: Union[RendererHooks, None] = None,
```
Comment (alexcjohnson) [inline dash/dash.py:378]: I think w should still mention this in the docstring, even though it's primarily meant for internal use. Something like "a dict of javascript functions for modifying behavior of the Dash renderer"

File: dash/_utils.py [hunk 3]
```diff
@@ -267,3 +269,15 @@ def coerce_to_list(obj):
...
+
+
+def hooks_to_js_object(hooks: Union[RendererHooks, None]) -> str:
+    if hooks is None:
+        return ""
+    hook_str = reduce(
+        lambda reduced, hook: f"{reduced}{hook[0]}: {hook[1]},",
+        hooks.items(),
+        "",
+    )
```
Comment (alexcjohnson) [inline dash/_utils.py:281]: heh writing Python like it's JS 😎 
```suggestion
    hook_str = ",".join(f"{key}: {val}" for key, val in hooks.items())
```

**Review by alexcjohnson — APPROVED** [review]
💃 Looks good, just had a couple of minor comments!

File: dash/dash.py (hunk 2 above)
Comment (KoolADE85) [reply dash/dash.py:382]: Good call, done!

File: dash/_utils.py (hunk 3 above)
Comment (KoolADE85) [reply dash/_utils.py]: Spot my native language :) 
Addressed in 5e3688b

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,13 @@ This project adheres to [Semantic Versioning](https://semver.org/).
 
 - [#2610](https://github.com/plotly/dash/pull/2610) Load plotly.js bundle/version from plotly.py
 
+## UNRELEASED
+
```
Comment (alexcjohnson) [inline CHANGELOG.md:11]:
```suggestion
```
(cruft from updating the branch via GUI)

**Pull request merged**